### DISCUSSION TITLE
Cannot change `primaryColor` based on Dark Mode

### DISCUSSION THREAD

**Discussion by lgaspari** [mantinedev/mantine#4829] [Q&A]
Hi, I'm slowly migrating from **v6 to v7** when I stumble upon this issue where I cannot use the `colorScheme` as it was before to render a different `primaryColor` based on it. Instead, it looks like I may need a workaround, but can't get my head around it.

I've read a couple of discussions about this that requires to use a nested `MantineProvider`, but doesn't seem like a good solution for my use case since it's not just a part of the application but all of it.

Is this even possible to achieve with the new version? I'm using Next.js 13.x and some pages are SSR, I think it's important to figure out a solution.

Thanks in advance!

```js
// v6
// theme.js (reduced file)
export function getTheme({ colorScheme }) {
  const isDarkMode = colorScheme === 'dark';

  const theme = createTheme({
    primaryColor: isDarkMode ? accentColorName : primaryColorName,
  });

  return theme;
}

// v6
// _app.js (reduced file)
import { COLOR_SCHEME_KEY } from '@common/app/constants/keys';
import { getTheme } from '@config/app/styles/theme';
import { ColorSchemeProvider, MantineProvider } from '@mantine/core';
import { Notifications } from '@mantine/notifications';
import { setCookie } from 'cookies-next';
import { useState } from 'react';

export default function App({ colorScheme, Component, pageProps }) {
  const [internalColorScheme, setInternalColorScheme] = useState(colorScheme);

  const toggleColorScheme = (value = undefined) => {
    const nextColorScheme =
      value || (internalColorScheme === 'dark' ? 'light' : 'dark');

    setInternalColorScheme(nextColorScheme);
    setCookie(COLOR_SCHEME_KEY, nextColorScheme, {
      maxAge: 60 * 60 * 24 * 30, // 1 month
    });
  };

  return (
    <ColorSchemeProvider
      colorScheme={internalColorScheme}
      toggleColorScheme={toggleColorScheme}
    >
      <MantineProvider
        theme={getTheme({ colorScheme: internalColorScheme })}
        withGlobalStyles
        withNormalizeCSS
      >
        <Notifications limit={5} />
        <Component {...pageProps} />
      </MantineProvider>
    </ColorSchemeProvider>
  );
}
```

**Comment by rtivital** [comment]
If you want to fully control color sheme value:
- Set `data-mantine-color-scheme` attribute on html element
- Create color scheme manager that will always return your value and pass it to MantineProvider – https://mantine.dev/theming/color-schemes/#color-scheme-manager
- Do not include ColorSchemeScript in your application

This way you can fully control the value on your side

**Reply by lgaspari** [reply]
Sorry but I don't think I understood what you're referring to. I don't really want to fully control the color scheme value, instead I'd only like to change theme's `primaryColor` based on color scheme (same as before).

Previously in v6, I'd manage the ColorScheme by using the `ColorSchemeProvider`, however the way it works right now doesn't allow me to get the `colorScheme` value without getting an error because I'm trying to get it before the `MantineProvider` is rendered.

Using your approach in v7, I'm not being able to get the `colorScheme` value after subscribing into the `colorSchemeManager` events.

```ts
import { localStorageColorSchemeManager, MantineProvider } from '@mantine/core';
import Head from 'next/head';
import { useEffect } from 'react';
import { theme } from '../theme';
import '@mantine/core/styles.css';

const colorSchemeManager = localStorageColorSchemeManager({
  key: 'my-color-scheme-manager-key',
});

const getTheme = ({ colorScheme }: { colorScheme: string }) => {
  console.log(colorScheme === 'dark' ? 'teal' : 'blue');
};

export default function App({ Component, pageProps }: any) {
  useEffect(() => {
    colorSchemeManager.subscribe((colorScheme) => {
      // not working
      console.log(colorScheme);
      console.log(getTheme({ colorScheme }));
    });

    return () => {
      colorSchemeManager.unsubscribe();
    };
  }, []);

  return (
    <MantineProvider theme={theme}>
      <Head>
        <title>Mantine Template</title>
        <meta
          name="viewport"
          content="minimum-scale=1, initial-scale=1, width=device-width, user-scalable=no"
        />
        <link rel="shortcut icon" href="/favicon.svg" />
      </Head>
      <Component {...pageProps} />
    </MantineProvider>
  );
}
```

**Reply by rtivital** [reply]
> I don't really want to fully control the color scheme value

It is not possible to implement the same logic without controlling the color scheme value because you need to access it before MantineProvider is initialized. In your v6 example, the color scheme is controlled. 

If the first approach is too complicated, you can use `forceColorScheme` with your color scheme value, then change theme based on this value. 

Another thing that you can try is to override CSS variables with resolver – https://mantine.dev/styles/css-variables/#css-variables-resolver
You can find CSS variables list here – https://mantine.dev/styles/css-variables-list/

**Reply by lgaspari** [reply]
Gotcha, yeah I know it's controlled I was just trying to point the fact that if I could avoid controlling it, that would be awesome to simplify things a little. Were you able to have the `subscribe()` working? The only way I can think of to maintain the `colorScheme` synced, unless I'm missing something.

For the CSS variables resolver I'd need to override all `--mantine-primary-color-{X}` variables, right?

**Reply by lgaspari** [reply]
Update: it looks like the subscription works only when changing its value from storage directly, but not if we use `useMantineColorScheme` as the following. Maybe I shouldn't toggle using that function? I'm kinda lost at this time :joy:

```ts
import { Button, Group, useMantineColorScheme } from '@mantine/core';

export default function IndexPage() {
  const { toggleColorScheme } = useMantineColorScheme();

  return (
    <Group mt={50} justify="center">
      <Button size="xl" onClick={toggleColorScheme}>
        Welcome to Mantine!
      </Button>
    </Group>
  );
}
```

Here's the `_app.js` code if you'd like to give it a try (I couldn't find the Mantine Issue Template for v7):

```ts
import {
  createTheme,
  localStorageColorSchemeManager,
  MantineProvider,
} from '@mantine/core';
import Head from 'next/head';
import { useEffect, useState } from 'react';
import '@mantine/core/styles.css';

const colorSchemeManager = localStorageColorSchemeManager();

const getTheme = ({ colorScheme }: { colorScheme: string }) => {
  return createTheme({
    primaryColor: colorScheme === 'dark' ? 'teal' : 'blue',
  });
};

export default function App({ Component, pageProps }: any) {
  const [internalColorScheme, setInternalColorScheme] = useState('light');

  useEffect(() => {
    colorSchemeManager.subscribe((colorScheme) => {
      console.log('subscription works!'); // currently only works when changing the value from storage manually
      setInternalColorScheme(colorScheme);
    });

    return () => {
      colorSchemeManager.unsubscribe();
    };
  }, []);

  return (
    <MantineProvider
      colorSchemeManager={colorSchemeManager}
      theme={getTheme({ colorScheme: internalColorScheme })}
    >
      <Head>
        <title>Mantine Template</title>
        <meta
          name="viewport"
          content="minimum-scale=1, initial-scale=1, width=device-width, user-scalable=no"
        />
        <link rel="shortcut icon" href="/favicon.svg" />
      </Head>
      <Component {...pageProps} />
    </MantineProvider>
  );
}
```
